### PR TITLE
Add tests for servlet-rw BufferedReader and PrintWriter instrumentations

### PR DESCRIPTION
**Pull request by shashank11p**
resolves https://github.com/hypertrace/javaagent/issues/240

### CODE REVIEW

File: instrumentation/servlet/servlet-rw/src/test/java/io/opentelemetry/javaagent/instrumentation/hypertrace/servlet/rw/reader/BufferedReaderInstrumentationTest.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+import java.io.CharArrayReader;
+import java.io.IOException;
+import org.ServletReadWriteContextAccess;
+import org.TestBufferedReader;
+import org.hypertrace.agent.core.instrumentation.buffer.*;
+import org.hypertrace.agent.testing.AbstractInstrumenterTest;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+public class BufferedReaderInstrumentationTest extends AbstractInstrumenterTest {
```
Comment (pavolloffay) [inline instrumentation/servlet/servlet-rw/src/test/java/io/opentelemetry/javaagent/instrumentation/hypertrace/servlet/rw/reader/BufferedReaderInstrumentationTest.java:30]: Could you please add tests for all methods that read data on the BufferedReader? The same for `PrintWriter`

File: instrumentation/servlet/servlet-rw/src/test/java/org/ServletReadWriteContextAccess.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+package org;
+
+import java.io.BufferedReader;
+import java.io.PrintWriter;
+import org.hypertrace.agent.core.instrumentation.buffer.BoundedCharArrayWriter;
+import org.hypertrace.agent.core.instrumentation.buffer.CharBufferSpanPair;
+
+public class ServletReadWriteContextAccess {
+
```
Comment (pavolloffay) [inline instrumentation/servlet/servlet-rw/src/test/java/org/ServletReadWriteContextAccess.java:25]: add a private constructor

File: instrumentation/servlet/servlet-rw/src/test/java/org/ServletReadWriteContextAccess.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+ */
+
+package org;
+
+import java.io.BufferedReader;
+import java.io.PrintWriter;
+import org.hypertrace.agent.core.instrumentation.buffer.BoundedCharArrayWriter;
+import org.hypertrace.agent.core.instrumentation.buffer.CharBufferSpanPair;
+
+public class ServletReadWriteContextAccess {
```
Comment (pavolloffay) [inline instrumentation/servlet/servlet-rw/src/test/java/org/ServletReadWriteContextAccess.java:24]: Rename to `BufferedReaderPrintWriterContextAccess`

File: instrumentation/servlet/servlet-rw/src/test/java/org/TestBufferedReader.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+ */
+
+package org;
+
+import java.io.BufferedReader;
+import java.io.Reader;
+import org.jetbrains.annotations.NotNull;
+
+public class TestBufferedReader extends BufferedReader {
+  public TestBufferedReader(@NotNull Reader in) {
```
Comment (pavolloffay) [inline instrumentation/servlet/servlet-rw/src/test/java/org/TestBufferedReader.java:24]: remove the annotation

File: instrumentation/servlet/servlet-rw/src/test/java/org/TestPrintWriter.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+ */
+
+package org;
+
+import java.io.PrintWriter;
+import java.io.Writer;
+import org.jetbrains.annotations.NotNull;
+
+public class TestPrintWriter extends PrintWriter {
+  public TestPrintWriter(@NotNull Writer out) {
```
Comment (pavolloffay) [inline instrumentation/servlet/servlet-rw/src/test/java/org/TestPrintWriter.java:24]: remove the annotation

File: instrumentation/servlet/servlet-rw/src/test/java/io/opentelemetry/javaagent/instrumentation/hypertrace/servlet/rw/writer/PrintWriterInstrumentationTest.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+public class PrintWriterInstrumentationTest extends AbstractInstrumenterTest {
+
+  private static final String BODY = "boobar";
+
+  @Test
+  public void write_single_char() {
+    PrintWriter printWriter = new TestPrintWriter(new CharArrayWriter());
```
Comment (pavolloffay) [inline instrumentation/servlet/servlet-rw/src/test/java/io/opentelemetry/javaagent/instrumentation/hypertrace/servlet/rw/writer/PrintWriterInstrumentationTest.java:35]: why do we need this class?

Comment (pavolloffay) [reply instrumentation/servlet/servlet-rw/src/test/java/io/opentelemetry/javaagent/instrumentation/hypertrace/servlet/rw/writer/PrintWriterInstrumentationTest.java:35]: `TestPrintWriter`

File: instrumentation/servlet/servlet-rw/src/test/java/io/opentelemetry/javaagent/instrumentation/hypertrace/servlet/rw/reader/BufferedReaderInstrumentationTest.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+  private static final String TEST_SPAN_NAME = "foo";
+  private static final String BODY = "boobar";
+
+  @Test
+  public void read() throws IOException {
+    Span span = TEST_TRACER.spanBuilder(TEST_SPAN_NAME).startSpan();
+
+    BufferedReader bufferedReader = new TestBufferedReader(new CharArrayReader(BODY.toCharArray()));
```
Comment (pavolloffay) [inline instrumentation/servlet/servlet-rw/src/test/java/io/opentelemetry/javaagent/instrumentation/hypertrace/servlet/rw/reader/BufferedReaderInstrumentationTest.java:39]: why do we need `TestBufferedReader`?

File: instrumentation/servlet/servlet-rw/src/test/java/io/opentelemetry/javaagent/instrumentation/hypertrace/servlet/rw/writer/PrintWriterInstrumentationTest.java (hunk 6 above)
Comment (shashank11p) [reply instrumentation/servlet/servlet-rw/src/test/java/io/opentelemetry/javaagent/instrumentation/hypertrace/servlet/rw/writer/PrintWriterInstrumentationTest.java:35]: We do no need this class. It is just to test if the instrumentation works for subclasses of PrintWriter as well

**Review by pavolloffay — APPROVED** [review]

**Pull request merged**